africanus/calibration/utils/compute_and_corrupt_vis.py

```diff
@@ -0,0 +1,153 @@
+from .utils import check_type
+from africanus.constants import minus_two_pi_over_c as m2pioc
+
+DIAG_DIAG = 0
+DIAG = 1
+FULL = 2
+
+
+def jones_mul_factory(mode):
+    if mode == DIAG_DIAG:
+        def jones_mul(a1j, model, a2j, uvw, freq, lm, out):
+            n_dir = np.shape(model)[0]
+            u, v, w = uvw
+            for s in range(n_dir):
+                l, m = lm[s]
+                n = np.sqrt(1 - l**2 - m**2)
+                real_phase = m2pioc * freq * (u*l + v*m + w*(n-1))
+                source_vis = model[s] * np.exp(1.0j*real_phase)/n
+                for c in range(out.shape[-1]):
+                    out[c] += a1j[s, c]*source_vis[c]*np.conj(a2j[s, c])
+    elif mode == DIAG:
+        def jones_mul(a1j, model, a2j, uvw, freq, lm, out):
+            n_dir = np.shape(model)[0]
+            u, v, w = uvw
+            for s in range(n_dir):
+                l, m = lm[s]
+                n = np.sqrt(1 - l**2 - m**2)
+                real_phase = m2pioc * freq * (u*l + v*m + w*(n-1))
+                source_vis = model[s] * np.exp(1.0j*real_phase)/n
+                out[0, 0] += a1j[s, 0]*source_vis[0, 0] * np.conj(a2j[s, 0])
+                out[0, 1] += a1j[s, 0]*source_vis[0, 1] * np.conj(a2j[s, 1])
+                out[1, 0] += a1j[s, 1]*source_vis[1, 0] * np.conj(a2j[s, 0])
+                out[1, 1] += a1j[s, 1]*source_vis[1, 1] * np.conj(a2j[s, 1])
+    elif mode == FULL:
+        def jones_mul(a1j, model, a2j, uvw, freq, lm, out):
+            n_dir = np.shape(model)[0]
+            u, v, w = uvw
+            for s in range(n_dir):
+                l, m = lm[s]
+                n = np.sqrt(1 - l**2 - m**2)
+                real_phase = m2pioc * freq * (u*l + v*m + w*(n-1))
+                source_vis = model[s] * np.exp(1.0j*real_phase)/n
+                # precompute resuable terms
+                t1 = a1j[s, 0, 0]*source_vis[0, 0]
+                t2 = a1j[s, 0, 1]*source_vis[1, 0]
+                t3 = a1j[s, 0, 0]*source_vis[0, 1]
+                t4 = a1j[s, 0, 1]*source_vis[1, 1]
+                tmp = np.conj(a2j[s].T)
+                # overwrite with result
+                out[0, 0] += t1*tmp[0, 0] +\
+                    t2*tmp[0, 0] +\
+                    t3*tmp[1, 0] +\
+                    t4*tmp[1, 0]
+                out[0, 1] += t1*tmp[0, 1] +\
+                    t2*tmp[0, 1] +\
+                    t3*tmp[1, 1] +\
+                    t4*tmp[1, 1]
+                t1 = a1j[s, 1, 0]*source_vis[0, 0]
+                t2 = a1j[s, 1, 1]*source_vis[1, 0]
+                t3 = a1j[s, 1, 0]*source_vis[0, 1]
+                t4 = a1j[s, 1, 1]*source_vis[1, 1]
+                out[1, 0] += t1*tmp[0, 0] +\
+                    t2*tmp[0, 0] +\
+                    t3*tmp[1, 0] +\
+                    t4*tmp[1, 0]
+                out[1, 1] += t1*tmp[0, 1] +\
+                    t2*tmp[0, 1] +\
+                    t3*tmp[1, 1] +\
+                    t4*tmp[1, 1]
+
+    return njit(nogil=True)(jones_mul)
```

Comment:
This may be a candidate for inlining

```suggestion
    return njit(nogil=True, inline='always')(jones_mul)
```